pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -36,260 +36,107 @@ var _ apis.Validatable = (*Pipeline)(nil)
 // taskRef or taskSpec, and in case of a pipeline task with taskRef, it has a reference to a valid task (task name)
 func validatePipelineTasks(ctx context.Context, tasks []PipelineTask, finalTasks []PipelineTask) *apis.FieldError {
 	// Names cannot be duplicated
 	taskNames := sets.NewString()
-	var err *apis.FieldError
+	var errs *apis.FieldError
 	for i, t := range tasks {
-		if err = validatePipelineTaskName(ctx, "spec.tasks", i, t, taskNames); err != nil {
-			return err
-		}
+		errs = errs.Also(validatePipelineTask(ctx, t, taskNames).ViaFieldIndex("tasks", i))
 	}
 	for i, t := range finalTasks {
-		if err = validatePipelineTaskName(ctx, "spec.finally", i, t, taskNames); err != nil {
-			return err
-		}
+		errs = errs.Also(validatePipelineTask(ctx, t, taskNames).ViaFieldIndex("finally", i))
 	}
-	return nil
+	return errs
 }
 
-func validatePipelineTaskName(ctx context.Context, prefix string, i int, t PipelineTask, taskNames sets.String) *apis.FieldError {
-	if errs := validation.IsDNS1123Label(t.Name); len(errs) > 0 {
+func validatePipelineTaskName(name string) *apis.FieldError {
+	if err := validation.IsDNS1123Label(name); len(err) > 0 {
 		return &apis.FieldError{
-			Message: fmt.Sprintf("invalid value %q", t.Name),
-			Paths:   []string{fmt.Sprintf(prefix+"[%d].name", i)},
+			Message: fmt.Sprintf("invalid value %q", name),
+			Paths:   []string{"name"},
 			Details: "Pipeline Task name must be a valid DNS Label." +
 				"For more info refer to https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names",
 		}
 	}
+	return nil
+}
+
+func validatePipelineTask(ctx context.Context, t PipelineTask, taskNames sets.String) *apis.FieldError {
+	errs := validatePipelineTaskName(t.Name)
 	// can't have both taskRef and taskSpec at the same time
 	if (t.TaskRef != nil && t.TaskRef.Name != "") && t.TaskSpec != nil {
-		return apis.ErrMultipleOneOf(fmt.Sprintf(prefix+"[%d].taskRef", i), fmt.Sprintf(prefix+"[%d].taskSpec", i))
+		errs = errs.Also(apis.ErrMultipleOneOf("taskRef", "taskSpec"))
 	}
 	// Check that one of TaskRef and TaskSpec is present
 	if (t.TaskRef == nil || (t.TaskRef != nil && t.TaskRef.Name == "")) && t.TaskSpec == nil {
-		return apis.ErrMissingOneOf(fmt.Sprintf(prefix+"[%d].taskRef", i), fmt.Sprintf(prefix+"[%d].taskSpec", i))
+		errs = errs.Also(apis.ErrMissingOneOf("taskRef", "taskSpec"))
 	}
 	// Validate TaskSpec if it's present
 	if t.TaskSpec != nil {
-		if err := t.TaskSpec.Validate(ctx); err != nil {
-			return err
-		}
+		errs = errs.Also(t.TaskSpec.Validate(ctx).ViaField("taskSpec"))
 	}
 	if t.TaskRef != nil && t.TaskRef.Name != "" {
-		// Task names are appended to the container name, which must exist and
-		// must be a valid k8s name
-		if errSlice := validation.IsQualifiedName(t.Name); len(errSlice) != 0 {
-			return apis.ErrInvalidValue(strings.Join(errSlice, ","), fmt.Sprintf(prefix+"[%d].name", i))
-		}
```

Comment:
This is less strict than `validatePipelineTaskName` so we don't really need to do it, at all